### PR TITLE
Split large requests to ElasticSearch

### PR DESCRIPTION
**Pull request by joehoyle**
ElasticPress doesn't have support for limiting request / posts indexed based off  a request size limit. We hook into ElasticPress every time it will make a request to check the payload size, and split it into multiple requests if needed.

### CODE REVIEW

File: inc/namespace.php [hunk 1]
```diff
@@ -1628,3 +1632,99 @@ function custom_search_results_post_type_args( array $args, string $post_type )
...
 }
+
+/**
+ * Split elasticsearch requests that will be too large for the request limit.
+ *
+ * ElasticPress doesn't have support for limiting request / posts indexed based off
+ * a request size limit. We hook into ElasticPress every time it will make a request
+ * to check the payload size, and split it into multiple requests if needed.
+ *
+ *
```
Comment (roborourke) [inline inc/namespace.php:1643]:
```suggestion
```

File: inc/namespace.php [hunk 2]
```diff
@@ -1628,3 +1632,99 @@ function custom_search_results_post_type_args( array $args, string $post_type )
...
+ * @param integer $failures       The current number of failures for this request.
+ * @return WP_Error|array         The overiden request.
+ */
+function split_large_ep_request( $request, array $query, array $args, int $failures ) {
+
+	if ( empty( $args['body'] ) || $args['method'] !== 'POST' || ! strpos( $query['url'], '/_bulk' ) ) {
+		return wp_remote_request( $query['url'], $args );
+	}
+
+	$body_size_limit = apply_filters( 'altis.search.request-size-limit', 1024 * 1024 * 9 ); // 9 MB
```
Comment (roborourke) [inline inc/namespace.php:1656]:
```suggestion
	// phpcs:ignore WordPress.NamingConventions.ValidHookName.UseUnderscores
	$body_size_limit = apply_filters( 'altis.search.request-size-limit', 1024 * 1024 * 9 ); // 9 MB
```

File: inc/namespace.php [hunk 3]
```diff
@@ -1628,3 +1632,99 @@ function custom_search_results_post_type_args( array $args, string $post_type )
...
+		// If adding the next chunk would take the current chunk over the limit, send the request now.
+		if ( strlen( $args['body'] ) + $next_chunk_position > $body_size_limit ) {
+			$requests[] = wp_remote_request( $query['url'], $args );
+			$args['body'] = '';
+		}
+
+		$args['body'] .= substr( $body, 0, $next_chunk_position );
+		$body = substr( $body, $next_chunk_position );
+
+		// Final request
```
Comment (roborourke) [inline inc/namespace.php:1686]:
```suggestion
		// Final request.
```

**Review by roborourke — APPROVED** [review]
I didn't realise Elasticsearch could support partial requests like this! That's pretty cool.

Trust everything is working ok? Feels a little risky to backport but I guess it's a bug essentially.

File: inc/namespace.php [hunk 4]
```diff
@@ -1628,3 +1632,99 @@ function custom_search_results_post_type_args( array $args, string $post_type )
...
+ * ElasticPress doesn't have support for limiting request / posts indexed based off
+ * a request size limit. We hook into ElasticPress every time it will make a request
+ * to check the payload size, and split it into multiple requests if needed.
+ *
+ *
+ * @param WP_Error|array $request The request to replace / override.
+ * @param array $query            The query to ElasticSearch.
+ * @param array $args             The request arguments.
+ * @param integer $failures       The current number of failures for this request.
+ * @return WP_Error|array         The overiden request.
```
Comment (roborourke) [inline inc/namespace.php:1648]:
```suggestion
 * @param array $query The query to ElasticSearch.
 * @param array $args The request arguments.
 * @param integer $failures The current number of failures for this request.
 * @return WP_Error|array The overriden request.
```

**Comment by joehoyle** [comment]
> I didn't realise Elasticsearch could support partial requests like this! That's pretty cool.

This specifically only works with the bulk request format, but yeah pretty handy.

> Trust everything is working ok? Feels a little risky to backport but I guess it's a bug essentially

Yeah, I'm a little nervous about back-porting immediately, maybe we could back-port a version that needs it enabled via the config or something?

**Comment by roborourke** [comment]
Yeah. In that case we could make preview releases for the versions as needed, I’d avoid adding config only for the backport.

**Pull request merged**